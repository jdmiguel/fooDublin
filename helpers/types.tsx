import { ReactNode } from 'react';

export type InitialAppState = {
  favorites: Restaurant[];
  relatedRestaurants: Restaurant[];
  breadcrumbs: BreadcrumbsData[];
};

// Core / ui components

export enum LogoSize {
  BIG = 'big',
  SMALL = 'small',
}

export enum LoaderType {
  CIRCLE = 'circle',
  LINE = 'line',
}

export enum LoaderMode {
  DARK = 'dark',
  LIGHT = 'light',
}

export type ListItem = {
  iconSrc?: string;
  id: number;
  name: string;
  path: string;
};

export enum CardType {
  SUGGESTION = 'suggestion',
  STANDART = 'standart',
  HIGHLIGHT = 'highlight',
}

export enum CustomLinkSize {
  BIG = 'big',
  SMALL = 'small',
}

// Restaurant data

export type Timming = {
  id: string;
  day: string;
  schedule: string;
};

export enum EntityType {
  CITY = 'city',
  SUBZONE = 'subzone',
}

export type RestaurantsRequestParams = {
  entity_id: number | null;
  cuisines: number | null;
  entity_type: EntityType.CITY | EntityType.SUBZONE;
  start?: number;
  sort?: string;
  order?: string;
  q?: string;
  count?: number;
};

export type Restaurant = {
  id: number;
  imgSrc: string;
  title: string;
  content: ReactNode | string;
  route: string;
  asRoute: string;
};

export type RawRestaurant = {
  restaurant: {
    R: {
      has_menu_status: {
        delivery: number;
        takeaway: number;
      };
      res_id: number;
      is_grocery_store: boolean;
    };
    apikey: string;
    id: number;
    name: string;
    url: string;
    location: {
      address: string;
      locality: string;
      city: string;
      city_id: number;
      latitude: number;
      longitude: number;
      zipcode: string;
      country_id: number;
      locality_verbose: string;
    };
    switch_to_order_menu: number;
    cuisines: string;
    timings: string;
    average_cost_for_two: number;
    price_range: number;
    currency: string;
    highlights: string[];
    offers: any[];
    opentable_support: number;
    is_zomato_book_res: number;
    mezzo_provider: string;
    is_book_form_web_view: number;
    book_form_web_view_url: string;
    book_again_url: string;
    thumb: string;
    user_rating: {
      aggregate_rating: number;
      rating_text: string;
      rating_color: string;
      rating_obj: {
        title: {
          text: number;
        };
        bg_color: {
          type: string;
          tint: number;
        };
      };
      votes: number;
    };
    all_reviews_count: number;
    photos_url: string;
    photo_count: number;
    menu_url: string;
    featured_image: string;
    has_online_delivery: number;
    is_delivering_now: number;
    store_type: string;
    include_bogo_offers: boolean;
    deeplink: string;
    is_table_reservation_supported: number;
    has_table_booking: number;
    events_url: string;
    phone_numbers: string;
    all_reviews: {
      reviews: [
        {
          review: any[];
        },
        {
          review: any[];
        },
        {
          review: any[];
        },
        {
          review: any[];
        },
        {
          review: any[];
        },
      ];
    };
    establishment: string[];
  };
};

export type RestaurantDetail = {
  imgSrc: string;
  thumbSrc: string;
  name: string;
  location: string;
  cuisines: string;
  timings: string;
  rating: number;
  votes: number;
  average: string;
  establishment: string;
  highlights: string[];
  phone: string;
  address: string;
};

export type Review = {
  id: number;
  userName: string;
  userImgSrc: string;
  rating: number;
  date: string;
  text: string;
};

export type RawReview = {
  review: {
    rating: number;
    review_text: string;
    id: number;
    rating_color: string;
    review_time_friendly: string;
    rating_text: string;
    timestamp: number;
    likes: number;
    user: {
      name: string;
      foodie_level: string;
      foodie_level_num: number;
      foodie_color: string;
      profile_url: string;
      profile_image: string;
      profile_deeplink: string;
    };
    comments_count: number;
  };
};

// Breadcrumbs

export enum BreadcrumbsType {
  HOME = 'home',
  SEARCH = 'search',
  DETAIL = 'detail',
  FAVORITES = 'favorite',
}

export type BreadcrumbsData = {
  text: string;
  route: string;
  asRoute: string;
  type:
    | BreadcrumbsType.HOME
    | BreadcrumbsType.SEARCH
    | BreadcrumbsType.DETAIL
    | BreadcrumbsType.FAVORITES;
};

// Actions

type AddFavoriteAction = {
  type: 'ADD_FAVORITE';
  favorite: Restaurant;
};

type DeleteFavoriteAction = {
  type: 'DELETE_FAVORITE';
  id: number;
};

type SetRelatedRestaurantsAction = {
  type: 'SET_RELATED_RESTAURANTS';
  relatedRestaurants: Restaurant[];
};

type ClearRelatedRestaurantsAction = {
  type: 'CLEAR_RELATED_RESTAURANTS';
};

type AddBreadcrumbsAction = {
  type: 'ADD_BREADCRUMBS';
  breadcrumbs: BreadcrumbsData;
};

type ReplaceBreadcrumbsAction = {
  type: 'REPLACE_BREADCRUMBS';
  index: number;
  breadcrumbs: BreadcrumbsData;
};

export type Actions =
  | AddFavoriteAction
  | DeleteFavoriteAction
  | SetRelatedRestaurantsAction
  | ClearRelatedRestaurantsAction
  | AddBreadcrumbsAction
  | ReplaceBreadcrumbsAction;
