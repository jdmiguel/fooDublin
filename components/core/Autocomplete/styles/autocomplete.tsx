import styled from 'styled-components';

export const StyledAutocomplete = styled.div<{ disabled: boolean }>`
  width: 100%;
  max-width: 550px;
  height: 55px;
  position: relative;
  cursor: pointer;
  pointer-events: ${({ disabled }) => disabled && 'none'};
`;

export const StyledInputWrapper = styled.div<{
  hasBorderBottomRadius: boolean;
}>`
  background-color: ${({ theme }) => theme.palette.LIGHT_MEDIUM};
  border-radius: 4px;
  transition: background-color 0.2s ease-out;
  &:hover {
    background-color: ${({ theme }) => theme.palette.PRIMARY_LIGHT};
  }

  ${({ hasBorderBottomRadius }) =>
    !hasBorderBottomRadius &&
    `border-bottom-left-radius: 0;
  border-bottom-right-radius: 0;`}
`;

export const StyledListboxWrapper = styled.div<{ isShowed: boolean }>`
  visibility: ${({ isShowed }) => (isShowed ? 'visible' : 'hidden')};
  opacity: ${({ isShowed }) => (isShowed ? '1' : '0')};
  transform: translateY(${({ isShowed }) => (isShowed ? '0' : '10px')});
  transition: opacity 0.2s ease 0s, transform 0.2s ease 0s;
  position: absolute;
  box-sizing: border-box;
  background: ${({ theme }) => theme.palette.LIGHT_MEDIUM};
  z-index: 2;
  top: 55px;
  left: 0;
  padding: 0;
  width: 100%;
  height: auto;
  max-height: 400px;
  overflow: auto;
  box-shadow: 0 2px 2px rgba(0, 0, 0, 0.35);
  border: 1px solid ${({ theme }) => theme.palette.LIGHT_MEDIUM};
  border-radius: 0 0 4px 4px;
  border-bottom-right-radius: 4px;
  border-bottom-left-radius: 4px;
  border-top: 0;
`;

export const StyledLoaderWrapper = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  width: 100%;
  height: 100%;
  background: ${({ theme }) => theme.palette.LIGHT_MEDIUM};
  @media only screen and (min-width: 992px) {
    min-height: 200px;
  }
`;

export const StyledListbox = styled.ul`
  width: 100%;
  -webkit-tap-highlight-color: transparent;
  @media only screen and (min-width: 992px) {
    min-height: 200px;
  }
`;

export const StyledNoSuggestionsWrapper = styled.div`
  height: 100%;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  @media only screen and (min-width: 992px) {
    min-height: 200px;
  }
`;

export const StyledErrorWrapper = styled.div`
  height: 100%;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  @media only screen and (min-width: 992px) {
    min-height: 200px;
  }
`;

export const StyledErrorButtonWrapper = styled.div`
  width: 150px;
  margin-top: 20px;
`;

export const StyledListboxItem = styled.li`
  width: 100%;
  border-bottom: 1px solid ${({ theme }) => theme.palette.LIGHT_MIN};
  list-style: none;
  display: flex;
  background-color: ${({ theme }) => theme.palette.LIGHT_MEDIUM};
  transition: background-color 0.2s ease-out;
  outline: none;
  -webkit-tap-highlight-color: transparent;
  cursor: pointer;
  &:hover {
    background-color: ${({ theme }) => theme.palette.PRIMARY_LIGHT};
  }
`;
