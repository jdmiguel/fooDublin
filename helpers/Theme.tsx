import React from 'react';
import { DefaultTheme, ThemeProvider } from 'styled-components';

export const theme: DefaultTheme = {
  palette: {
    PRIMARY: '#EC7E00',
    PRIMARY_DARK: '#D48C24',
    PRIMARY_MEDIUM: '#FBA52B',
    PRIMARY_LIGHT: '#FDF3E5',
    SECONDARY: '#660000',
    LIGHT_MAX: '#FFFFFF',
    LIGHT_MEDIUM: '#F3F3F3',
    LIGHT_MIN: '#E5EAED',
    LIGHT_SOFT: '#D8D8D8',
    DARK_MAX: '#222C3C',
    DARK_MEDIUM: '#585858',
    DARK_MIN: '#89959B',
    DARK_SOFT: '#A7A7A7',
  },
};

export const renderWithTheme = (component: any) => (
  <ThemeProvider theme={theme}>{component}</ThemeProvider>
);
