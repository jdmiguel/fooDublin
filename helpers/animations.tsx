import { keyframes, css } from 'styled-components';

const fadeIn = (endValue: number) => keyframes`
  0% { opacity: 0 }
  100% { opacity: ${endValue} }
`;

const fadeInFromPosY = (initialPosY: number) => keyframes`
  0% { opacity: 0; transform: translateY(${initialPosY}px); }
  100% { opacity: 1; ; transform: translateY(0); }
`;

const rotate = keyframes`
  0% { transform: rotate(0deg); }
  100% { transform: rotate(360deg); }
`;

const moveX = keyframes`
  0% { transform: translateX(-100%); }
  100% { transform: translateX(0); }
`;

export const fadeInAnimation = css<{ endValue?: number }>`
  animation: ${({ endValue }) => fadeIn(endValue || 1)} 0.15s linear;
`;

export const fadeInFromPosYAnimation = css<{
  initialPosY?: number;
  animationDuration?: number;
  animationDelay?: number;
}>`
  animation-name: ${({ initialPosY }) => fadeInFromPosY(initialPosY || 50)};
  animation-duration: ${({ animationDuration }) => animationDuration || 0.15}s;
  animation-delay: ${({ animationDelay }) => animationDelay || 0.25}s;
  animation-fill-mode: forwards;
  animation-timing-function: ease-out;
`;

export const circleLoaderAnimation = css`
  animation: ${rotate} 0.8s linear infinite;
`;

export const lineLoaderAnimation = css`
  animation: ${moveX} 1s ease-in-out infinite forwards;
`;
