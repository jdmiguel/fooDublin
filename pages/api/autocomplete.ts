import axios from 'axios';
import { NextApiRequest, NextApiResponse } from 'next';
import { BASE_API } from '@/store/statics';

export default async (req: NextApiRequest, res: NextApiResponse) => {
  const { searchText, latitude, longitude } = req.query;

  try {
    const response = await axios(`${BASE_API}autocomplete`, {
      method: 'GET',
      headers: {
        'Access-Control-Allow-Origin': '*',
        'x-requested-with': 'xmlhttprequest',
        accept: 'application/json',
        Authorization: `Bearer ${process.env.YELP_API_KEY}`,
      },
      params: {
        text: searchText,
        latitude,
        longitude,
      },
    });

    res.status(200).json(response.data.businesses);
  } catch (error: any) {
    res.status(error.response?.status || 500).json({ error: 'An error occurred' });
  }
};
