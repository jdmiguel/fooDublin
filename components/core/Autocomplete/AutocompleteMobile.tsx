import React, { useState, useEffect, useRef } from 'react';
import Link from 'next/link';
import styled from 'styled-components';
import { disableBodyScroll, enableBodyScroll } from 'body-scroll-lock';

import { Input } from '../Input/Input';
import { Card }, { CardType } from '../Card/Card';
import { Loader } from '../Loader/Loader';
import { BlockTitle } from '../BlockTitle/BlockTitle';

import {
  PlaceholderText,
  DEFAULT_TEXT_LOADING,
} from '../../../helpers/staticData';
import { Restaurant } from '../../../helpers/types';

export type AutocompleteMobileProps = {
  className?: string;
  hasSearchIcon?: boolean;
  suggestions: Restaurant[];
  loading: boolean;
  fetchSuggestions: (search: string) => void;
  selectSuggestion: (id: string, name: string) => void;
};

const StyledAutocompleteMobile = styled.div`
  width: 100%;
  max-width: 550px;
  height: 100%;
`;

const StyledLabel = styled.div`
  position: relative;
  border-radius: 4px;
  background-color: ${({ theme }) => theme.palette.LIGHT_MEDIUM};
`;

const StyledLabelButton = styled.button`
  width: 100%;
  height: 55px;
  background-color: ${({ theme }) => theme.palette.LIGHT_MEDIUM};
  line-height: 55px;
  padding: 0 4px 0 8px;
  display: flex;
  align-items: center;
  border-radius: 4px;
  cursor: pointer;
  outline: none;
  -webkit-tap-highlight-color: transparent;
  i {
    font-size: 1.2rem;
    margin-left: 5px;
    margin-right: 10px;
    color: ${({ theme }) => theme.palette.DARK_SOFT};
  }
  span {
    color: ${({ theme }) => theme.palette.DARK_SOFT};
  }
`;

const StyledModal = styled.div<{ isShowed: boolean }>`
  visibility: ${({ isShowed }) => (isShowed ? 'visible' : 'hidden')};
  opacity: ${({ isShowed }) => (isShowed ? '1' : '0')};
  transition: opacity 0.2s ease 0s;
  position: fixed;
  box-sizing: border-box;
  background: ${({ theme }) => theme.palette.LIGHT_MEDIUM};
  z-index: 2;
  max-width: 767px;
  top: 0;
  left: 0;
  padding: 20px 20px 10px;
  width: 100%;
  height: 100%;
  box-shadow: 0 2px 2px rgba(0, 0, 0, 0.35);
  border: 1px solid ${({ theme }) => theme.palette.LIGHT_MEDIUM};
  border-radius: 0 0 4px 4px;
  border-bottom-right-radius: 4px;
  border-bottom-left-radius: 4px;
  border-top: 0;
`;

const StyleHeading = styled.div`
  margin: 15px 0 25px;
  display: flex;
  justify-content: space-between;
`;

const StyleHeadingText = styled(BlockTitle)`
  font-size: 1.3rem;
`;

const StyleHeadingButton = styled.button`
  display: block;
  transition: opacity 0.2s ease-out;
  cursor: pointer;
  background-color: transparent;
  outline: none;
  -webkit-tap-highlight-color: transparent;
  i {
    font-size: 1.5rem;
    font-weight: 600;
    color: ${({ theme }) => theme.palette.DARK_MEDIUM};
  }
`;

const StyledInput = styled(Input)`
  border: 1px solid ${({ theme }) => theme.palette.LIGHT_SOFT};
  background-color: ${({ theme }) => theme.palette.LIGHT_MEDIUM};
  transition: background-color 0.2s ease-out;
  outline: none;
  &:hover {
    background-color: ${({ theme }) => theme.palette.PRIMARY_LIGHT};
  }
  &:focus {
    background-color: ${({ theme }) => theme.palette.PRIMARY_LIGHT};
  }

  input {
    cursor: pointer;
  }
`;

const StyledListboxWrapper = styled.div<{ isShowed: boolean }>`
  visibility: ${({ isShowed }) => (isShowed ? 'visible' : 'hidden')};
  opacity: ${({ isShowed }) => (isShowed ? '1' : '0')};
  transition: opacity 0.2s ease 0s;
  position: absolute;
  box-sizing: border-box;
  background: ${({ theme }) => theme.palette.LIGHT_MEDIUM};
  z-index: 1;
  top: 150px;
  left: 0;
  padding: 0 10px 173px;
  margin-top: 15px;
  width: 100%;
  height: 100%;
  max-height: 100vh;
  overflow: auto;
  border: 1px solid ${({ theme }) => theme.palette.LIGHT_MEDIUM};
  border-radius: 4px;
`;

const StyledLoader = styled(Loader)`
  width: 100%;
  height: 100%;
  background: ${({ theme }) => theme.palette.LIGHT_MEDIUM};
  box-sizing: border-box;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  @media only screen and (min-width: 992px) {
    min-height: 200px;
  }
`;

const StyledListbox = styled.ul`
  width: 100%;
`;

const StyledListboxItem = styled.li`
  width: 100%;
  border-bottom: 1px solid ${({ theme }) => theme.palette.LIGHT_MIN};
  list-style: none;
  overflow: hidden;
  background-color: ${({ theme }) => theme.palette.LIGHT_MEDIUM};
  display: flex;
  transition: background-color 0.2s ease-out;
  outline: none;
  color: ${({ theme }) => theme.palette.DARK_MAX};
  -webkit-tap-highlight-color: transparent;
  cursor: pointer;
  &:hover {
    background-color: ${({ theme }) => theme.palette.PRIMARY_LIGHT};
  }
`;

export const AutocompleteMobile: React.FC<AutocompleteMobileProps> = ({
  suggestions,
  fetchSuggestions,
  selectSuggestion,
  loading,
  hasSearchIcon,
  className,
}) => {
  const listboxWrapperRef = useRef<HTMLDivElement>(null);
  const isSuggestable = useRef(true);
  const [value, setValue] = useState('');
  const [focusedPlaceholder, setFocusedPlaceholder] = useState(
    PlaceholderText.BLURRED,
  );
  const [labelTextShowed, setLabelTextShowed] = useState(
    PlaceholderText.BLURRED as string,
  );
  const [isModalShowed, setIsModalShowed] = useState(false);
  const [isListboxFocused, setIsListboxFocused] = useState(false);

  useEffect(() => {
    if (isModalShowed && listboxWrapperRef.current) {
      disableBodyScroll(listboxWrapperRef.current);
    } else if (!isModalShowed && listboxWrapperRef.current) {
      setValue('');
      setIsListboxFocused(false);
      enableBodyScroll(listboxWrapperRef.current);
    }
  }, [isModalShowed]);

  useEffect(() => {
    if (value.length > 2 && isSuggestable.current) {
      fetchSuggestions(value);
    }
  }, [value]);

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const currentValue = event.target.value;

    setIsListboxFocused(currentValue.length > 2);
    setValue(currentValue);
  };

  const handleInputFocus = () => {
    if (!value) {
      setFocusedPlaceholder(PlaceholderText.FOCUSED);
    } else if (value.length > 2) {
      setIsListboxFocused(true);
    }
  };

  const handleInputBlur = () => {
    !value && setFocusedPlaceholder(PlaceholderText.BLURRED);
  };

  const handleSuggestionClick = (restaurantId: string, showedText: string) => {
    setValue(showedText);
    setLabelTextShowed(showedText);
    setIsModalShowed(false);
    isSuggestable.current = false;
    selectSuggestion(restaurantId, showedText);
  };

  const handleCloseModal = () => {
    setIsModalShowed(false);
    setIsListboxFocused(false);
  };

  return (
    <StyledAutocompleteMobile data-testid="autocomplete" className={className}>
      <>
        <StyledLabel data-testid="label">
          <StyledLabelButton
            type="button"
            onClick={() => setIsModalShowed(true)}
          >
            <i data-testid={'label-icon'} className="material-icons">
              search
            </i>
            <span>{labelTextShowed}</span>
          </StyledLabelButton>
        </StyledLabel>
        <StyledModal isShowed={isModalShowed} data-testid="modal">
          <StyleHeading>
            <StyleHeadingText text="Dublin restaurants" />
            <StyleHeadingButton type="button" onClick={handleCloseModal}>
              <i className="material-icons">close</i>
            </StyleHeadingButton>
          </StyleHeading>
          <StyledInput
            type="text"
            placeholder={focusedPlaceholder}
            onChange={handleChange}
            value={value}
            onFocus={handleInputFocus}
            onBlur={handleInputBlur}
            hasSearchIcon={hasSearchIcon}
          />
          <StyledListboxWrapper
            isShowed={isListboxFocused && suggestions.length > 0}
            data-testid="listbox-wrapper"
            ref={listboxWrapperRef}
          >
            {loading ? (
              <StyledLoader text={DEFAULT_TEXT_LOADING} />
            ) : (
              <StyledListbox role="listbox">
                {suggestions.map(
                  ({
                    id,
                    imgSrc,
                    title,
                    content,
                    route,
                    asRoute,
                  }: Restaurant) => (
                    <StyledListboxItem key={id} role="option">
                      <Link href={route} as={asRoute}>
                        <Card
                          imgSrc={imgSrc}
                          title={title}
                          content={content}
                          onClick={() => handleSuggestionClick(id, title)}
                          type={CardType.SUGGESTION}
                        />
                      </Link>
                    </StyledListboxItem>
                  ),
                )}
              </StyledListbox>
            )}
          </StyledListboxWrapper>
        </StyledModal>
      </>
    </StyledAutocompleteMobile>
  );
};
