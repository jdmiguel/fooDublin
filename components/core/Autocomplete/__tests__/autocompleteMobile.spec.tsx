import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import 'jest-styled-components';

import { AutocompleteMobile } from '../AutocompleteMobile';

import { AUTOCOMPLETE_PROPS_MOCK } from '../__mocks__/autocomplete.mocks';

import { renderWithTheme } from '../../../../helpers/Theme';

describe('Component: AutocompleteMobile', () => {
  it('should render', () => {
    const { container } = render(
      renderWithTheme(<AutocompleteMobile {...AUTOCOMPLETE_PROPS_MOCK} />),
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should show / hide modal', () => {
    const { getByTestId } = render(
      renderWithTheme(<AutocompleteMobile {...AUTOCOMPLETE_PROPS_MOCK} />),
    );
    const labelButton = getByTestId('label').querySelector('button');
    const modal = getByTestId('modal');
    const closeButton = modal.querySelector('button');

    // initiality, modal has not been showed
    expect(modal).toHaveStyleRule('opacity', '0');
    expect(modal).toHaveStyleRule('visibility', 'hidden');

    // show modal by clicking label button
    fireEvent.click(labelButton);
    expect(modal).toHaveStyleRule('opacity', '1');
    expect(modal).toHaveStyleRule('visibility', 'visible');

    // hide modal by clicking close button
    fireEvent.click(closeButton);
    expect(modal).toHaveStyleRule('opacity', '0');
    expect(modal).toHaveStyleRule('visibility', 'hidden');
  });

  it('should show suggestions list with a search of 3 characters and call callback function', () => {
    const handleFetchSuggestion = jest.fn();
    const { getByTestId, getByPlaceholderText } = render(
      renderWithTheme(
        <AutocompleteMobile
          {...AUTOCOMPLETE_PROPS_MOCK}
          fetchSuggestions={handleFetchSuggestion}
        />,
      ),
    );
    const listboxWrapper = getByTestId('listbox-wrapper');

    // check if suggestions list is not showed
    expect(listboxWrapper).toHaveStyleRule('opacity', '0');
    expect(listboxWrapper).toHaveStyleRule('visibility', 'hidden');

    // text 3 characters, check if suggestions list is showed and call callback function
    fireEvent.change(getByPlaceholderText('Search for locals...'), {
      target: { value: 'tre' },
    });

    expect(listboxWrapper).toHaveStyleRule('opacity', '1');
    expect(listboxWrapper).toHaveStyleRule('visibility', 'visible');
    expect(handleFetchSuggestion).toHaveBeenCalled();
  });

  it('should show/hide suggestions list', () => {
    const handleSelectSuggestion = jest.fn();
    const { getByTestId, getByPlaceholderText } = render(
      renderWithTheme(
        <AutocompleteMobile
          {...AUTOCOMPLETE_PROPS_MOCK}
          selectSuggestion={handleSelectSuggestion}
        />,
      ),
    );
    const closeButton = getByTestId('modal').querySelector('button');
    const listboxWrapper = getByTestId('listbox-wrapper');
    const input = getByPlaceholderText('Search for locals...');

    // show suggestions list by texting three characters
    fireEvent.change(input, { target: { value: 'tre' } });

    expect(listboxWrapper).toHaveStyleRule('opacity', '1');
    expect(listboxWrapper).toHaveStyleRule('visibility', 'visible');

    // hide suggestions list by texting two characters
    fireEvent.change(input, { target: { value: 'tr' } });

    expect(listboxWrapper).toHaveStyleRule('opacity', '0');
    expect(listboxWrapper).toHaveStyleRule('visibility', 'hidden');

    // show suggestions list by texting three characters and hide it by clicking close button
    fireEvent.change(input, { target: { value: 'tre' } });
    fireEvent.click(closeButton);

    expect(listboxWrapper).toHaveStyleRule('opacity', '0');
    expect(listboxWrapper).toHaveStyleRule('visibility', 'hidden');

    // show suggestions list by texting three characters and call callback function by clicking any suggestion
    fireEvent.change(input, { target: { value: 'tre' } });
    const firstSuggestionLink = listboxWrapper.querySelectorAll('li')[0].querySelector('a');
    fireEvent.click(firstSuggestionLink);

    expect(handleSelectSuggestion).toHaveBeenCalled();
  });

  it('should be disabled', () => {
    const AUTOCOMPLETE_PROPS_MOCK_DISABLED = {
      ...AUTOCOMPLETE_PROPS_MOCK,
      disabled: true,
    };
    const { container } = render(
      renderWithTheme(<AutocompleteMobile {...AUTOCOMPLETE_PROPS_MOCK_DISABLED} />),
    );
    const autocomplete = container.firstChild as HTMLDivElement;

    expect(autocomplete).toHaveStyleRule('pointer-events', 'none');
  });
});
