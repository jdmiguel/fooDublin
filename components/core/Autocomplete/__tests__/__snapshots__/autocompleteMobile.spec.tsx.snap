// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: AutocompleteMobile should render 1`] = `
.c8 {
  width: 100%;
  border-radius: 4px;
  padding: 0 20px;
  background: #FFFFFF;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c8 i {
  font-size: 1.2rem;
  color: #A7A7A7;
  margin-right: 10px;
}

.c10 {
  width: 100%;
  height: 55px;
  padding: 10px 0;
  line-height: 20px;
  border: none;
  outline: none;
  color: #2f3235;
  background: transparent;
  cursor: pointer;
  outline: none;
  -webkit-tap-highlight-color: transparent;
}

.c5 {
  color: #585858;
  font-size: 1.2rem;
  font-weight: 500;
}

.c0 {
  width: 100%;
  max-width: 550px;
  height: 100%;
}

.c1 {
  position: relative;
  border-radius: 4px;
  background-color: #FFFFFF;
}

.c2 {
  width: 100%;
  height: 55px;
  background-color: #FFFFFF;
  line-height: 55px;
  padding: 0 4px 0 8px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 4px;
  cursor: pointer;
  outline: none;
  -webkit-tap-highlight-color: transparent;
}

.c2 i {
  font-size: 1.2rem;
  margin-left: 5px;
  margin-right: 10px;
  color: #A7A7A7;
}

.c2 span {
  color: #A7A7A7;
}

.c3 {
  visibility: hidden;
  opacity: 0;
  -webkit-transition: opacity 0.2s ease 0s;
  transition: opacity 0.2s ease 0s;
  position: fixed;
  box-sizing: border-box;
  background: #FFFFFF;
  z-index: 2;
  max-width: 767px;
  top: 0;
  left: 0;
  padding: 20px 20px 10px;
  width: 100%;
  height: 100%;
  box-shadow: 0 2px 2px rgba(0,0,0,0.35);
  border: 1px solid #FFFFFF;
  border-radius: 0 0 4px 4px;
  border-bottom-right-radius: 4px;
  border-bottom-left-radius: 4px;
  border-top: 0;
}

.c4 {
  margin: 15px 0 25px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c6 {
  font-size: 1.3rem;
}

.c7 {
  display: block;
  -webkit-transition: opacity 0.2s ease-out;
  transition: opacity 0.2s ease-out;
  cursor: pointer;
  background-color: transparent;
  outline: none;
  -webkit-tap-highlight-color: transparent;
}

.c7 i {
  font-size: 1.5rem;
  font-weight: 600;
  color: #585858;
}

.c9 {
  border: 1px solid #D8D8D8;
  background-color: #FFFFFF;
  -webkit-transition: background-color 0.2s ease-out;
  transition: background-color 0.2s ease-out;
  outline: none;
}

.c9:hover {
  background-color: #FDF3E5;
}

.c9:focus {
  background-color: #FDF3E5;
}

.c9 input {
  cursor: pointer;
}

.c11 {
  visibility: hidden;
  opacity: 0;
  -webkit-transition: opacity 0.2s ease 0s;
  transition: opacity 0.2s ease 0s;
  position: absolute;
  box-sizing: border-box;
  background: #FFFFFF;
  z-index: 1;
  top: 150px;
  left: 0;
  padding: 0 10px 173px;
  margin-top: 15px;
  width: 100%;
  height: 100%;
  max-height: 100vh;
  overflow: auto;
  border: 1px solid #FFFFFF;
  border-radius: 4px;
}

.c12 {
  width: 100%;
}

.c13 {
  width: 100%;
  border-bottom: 1px solid #E5EAED;
  list-style: none;
  padding: 11px;
  overflow: hidden;
  background-color: #FFFFFF;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-transition: background-color 0.2s ease-out;
  transition: background-color 0.2s ease-out;
  outline: none;
  color: #2f3235;
  -webkit-tap-highlight-color: transparent;
  cursor: pointer;
}

.c13:hover {
  background-color: #FDF3E5;
}

.c14 {
  width: 30px;
  height: 30px;
  border: solid #FFFFFF 1px;
  display: inline-block;
  border-radius: 4px;
  -webkit-appearance: button-bevel;
  box-shadow: 1px 2px 5px #888;
}

.c15 {
  margin-left: 12px;
}

.c16 {
  text-align: left;
  color: #2f3235;
  font-size: 1.05rem;
}

.c17 {
  text-align: left;
  color: #A7A7A7;
  font-size: 0.85rem;
}

@media only screen and (min-width:768px) {
  .c8 {
    max-width: 550px;
  }
}

@media only screen and (min-width:992px) {
  .c5 {
    font-size: 1.45rem;
  }
}

<div
  class="c0"
  data-testid="autocomplete"
>
  <div
    class="c1"
    data-testid="label"
  >
    <button
      class="c2"
      type="button"
    >
      <i
        class="material-icons"
        data-testid="label-icon"
      >
        search
      </i>
      <span>
        Search for locals...
      </span>
    </button>
  </div>
  <div
    class="c3"
    data-testid="modal"
  >
    <div
      class="c4"
    >
      <h4
        class="c5 c6"
      >
        Dublin restaurants
      </h4>
      <button
        class="c7"
        type="button"
      >
        <i
          class="material-icons"
        >
          close
        </i>
      </button>
    </div>
    <div
      class="c8 c9"
      data-testid="input-wrapper"
    >
      <i
        class="material-icons"
        data-testid="input-icon"
      >
        search
      </i>
      <input
        class="c10"
        placeholder="Search for locals..."
        type="text"
        value=""
      />
    </div>
    <div
      class="c11"
      data-testid="listbox-wrapper"
    >
      <ul
        class="c12"
        role="listbox"
      >
        <li
          class="c13"
          role="option"
        >
          <div
            class="LazyImage-Placeholder"
          >
            <img
              alt="Manifesto"
              class="c14"
              src="http://localhost:3000/static/images/generic-thumb.png"
            />
          </div>
          <div
            class="c15"
          >
            <p
              class="c16"
            >
              Manifesto
            </p>
            <p
              class="c17"
            >
              Rathmines
            </p>
          </div>
        </li>
        <li
          class="c13"
          role="option"
        >
          <div
            class="LazyImage-Placeholder"
          >
            <img
              alt="Dragon Inn"
              class="c14"
              src="http://localhost:3000/static/images/generic-thumb.png"
            />
          </div>
          <div
            class="c15"
          >
            <p
              class="c16"
            >
              Dragon Inn
            </p>
            <p
              class="c17"
            >
              South City West
            </p>
          </div>
        </li>
        <li
          class="c13"
          role="option"
        >
          <div
            class="LazyImage-Placeholder"
          >
            <img
              alt="Casa Del Toro"
              class="c14"
              src="http://localhost:3000/static/images/generic-thumb.png"
            />
          </div>
          <div
            class="c15"
          >
            <p
              class="c16"
            >
              Casa Del Toro
            </p>
            <p
              class="c17"
            >
              Drumcondra
            </p>
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
